apps/chat/agent/openapi_schema.py

```diff
@@ -0,0 +1,194 @@
+from collections import defaultdict
+from typing import TypedDict
+
+from langchain_community.tools import APIOperation
+from langchain_community.utilities.openapi import OpenAPISpec
+from openapi_pydantic import DataType, Parameter, Reference, Schema
+from pydantic import BaseModel, Field, create_model
+
+
+class FunctionDef(TypedDict):
+    name: str
+    description: str
+    method: str
+    url: str
+    args_schema: type[BaseModel]
+
+
+def openapi_spec_op_to_function_def(spec: OpenAPISpec, path: str, method: str) -> FunctionDef:
```

Comment:
Where is this function being used?

Reply:
It isn't yet - I'm going to refactor the Tool now to use this instead of the current method.